File: backend/services/stigmer-server/pkg/domain/agentexecution/approval/lease_scope_corpus_test.go
package approval

import (
	"encoding/json"
	"os"
	"path/filepath"
	"runtime"
	"testing"

	agentexecutionv1 "github.com/stigmer/stigmer/apis/stubs/go/ai/stigmer/agentic/agentexecution/v1"
)

// leaseScopeVector mirrors one entry of apis/testdata/hitl/lease-scope/vectors.json.
// expected is null (no leasable scope), {"category": ...}, or {"server": ...}.
type leaseScopeVector struct {
	Name  string `json:"name"`
	Input struct {
		ToolName      string `json:"toolName"`
		McpServerSlug string `json:"mcpServerSlug"`
	} `json:"input"`
	Expected *struct {
		Category string `json:"category"`
		Server   string `json:"server"`
	} `json:"expected"`
}

// TestLeaseScopeCorpus is the OSS half of the cross-edition lease-scope parity
// gate: every vector in apis/testdata/hitl/lease-scope is derived by
// DeriveLeaseScope and must equal the expected scope. The runner (TS) and Cloud
// (Java) editions load the same file, so a drift fails one of the three suites.
func TestLeaseScopeCorpus(t *testing.T) {
	corpus := loadLeaseScopeCorpus(t)

	// Guard the guard: a silently empty corpus would pass for the wrong reason.
	if len(corpus) < 10 {
		t.Fatalf("loaded %d lease-scope vectors, want >= 10 (corpus missing or not discovered)", len(corpus))
	}

	for _, v := range corpus {
		t.Run(v.Name, func(t *testing.T) {
			tc := &agentexecutionv1.ToolCall{
				Name:          v.Input.ToolName,
				McpServerSlug: v.Input.McpServerSlug,
			}
			scope, ok := DeriveLeaseScope(tc)

			if v.Expected == nil {
				if ok {
					t.Fatalf("expected no leasable scope, got %+v", scope)
				}
				return
			}
			if !ok {
				t.Fatalf("expected scope %+v, got none", *v.Expected)
			}
			want := LeaseScope{Category: v.Expected.Category, Server: v.Expected.Server}
			if scope != want {
				t.Errorf("scope = %+v, want %+v", scope, want)
			}
		})
	}
}

func loadLeaseScopeCorpus(t *testing.T) []leaseScopeVector {
	t.Helper()
	_, thisFile, _, ok := runtime.Caller(0)
	if !ok {
		t.Fatal("runtime.Caller failed; cannot locate the lease-scope corpus")
	}
	// thisFile: backend/services/stigmer-server/pkg/domain/agentexecution/approval/lease_scope_corpus_test.go
	// repo root is seven directories up; the corpus lives under apis/testdata/hitl.
	repoRoot := filepath.Join(filepath.Dir(thisFile), "../../../../../../..")
	path := filepath.Join(repoRoot, "apis", "testdata", "hitl", "lease-scope", "vectors.json")

	raw, err := os.ReadFile(path)
	if err != nil {
		t.Fatalf("reading lease-scope corpus %s: %v", path, err)
	}
	var doc struct {
		Vectors []leaseScopeVector `json:"vectors"`
	}
	if err := json.Unmarshal(raw, &doc); err != nil {
		t.Fatalf("decoding lease-scope corpus %s: %v", path, err)
	}
	return doc.Vectors
}

File: backend/services/stigmer-server-ts/src/domain/agentexecution/approval/__tests__/lease-scope-corpus.test.ts
/**
 * The TS-server half of the cross-edition lease-scope parity gate
 * (apis/testdata/hitl/lease-scope) — ports lease_scope_corpus_test.go +
 * lease_scope_test.go: every vector derives through deriveLeaseScope and
 * must equal the expected scope. The runner (TS) and Cloud (Java)
 * editions load the same file, so a drift fails one of the suites.
 */
import path from "node:path";

import { create } from "@bufbuild/protobuf";
import { describe, expect, it } from "vitest";

import { ToolCallSchema } from "@stigmer/protos/ai/stigmer/agentic/agentexecution/v1/message_pb";

import { deriveLeaseScope, sameLeaseScope } from "../lease-scope.js";
import { hitlCorpusDir, readCorpusJson } from "./corpus-support.js";

interface LeaseScopeVector {
  name: string;
  input: { toolName?: string; mcpServerSlug?: string };
  expected: { category?: string; server?: string } | null;
}

describe("shared lease-scope corpus", () => {
  const doc = readCorpusJson(
    path.join(hitlCorpusDir(), "lease-scope", "vectors.json"),
  ) as unknown as { vectors: LeaseScopeVector[] };

  // Guard the guard (Go asserts >= 10 too).
  it("discovers the corpus", () => {
    expect(doc.vectors.length).toBeGreaterThanOrEqual(10);
  });

  for (const v of doc.vectors) {
    it(v.name, () => {
      const tc = create(ToolCallSchema, {
        name: v.input.toolName ?? "",
        mcpServerSlug: v.input.mcpServerSlug ?? "",
      });
      const scope = deriveLeaseScope(tc);

      if (v.expected === null) {
        expect(scope, "expected no leasable scope").toBeUndefined();
        return;
      }
      expect(scope, "expected a leasable scope").toBeDefined();
      expect(
        sameLeaseScope(scope as NonNullable<typeof scope>, {
          category: v.expected.category ?? "",
          server: v.expected.server ?? "",
        }),
      ).toBe(true);
    });
  }
});

describe("deriveLeaseScope unit pins (lease_scope_test.go)", () => {
  it("an MCP slug takes precedence over any category lookup", () => {
    const scope = deriveLeaseScope(
      create(ToolCallSchema, { name: "shell", mcpServerSlug: "github" }),
    );
    expect(scope).toEqual({ category: "", server: "github" });
  });

  it("a gated built-in derives its category", () => {
    const scope = deriveLeaseScope(create(ToolCallSchema, { name: "Write" }));
    expect(scope).toEqual({ category: "write", server: "" });
  });

  it("a read-only or unknown tool has no leasable scope", () => {
    expect(
      deriveLeaseScope(create(ToolCallSchema, { name: "read_file" })),
    ).toBeUndefined();
    expect(
      deriveLeaseScope(create(ToolCallSchema, { name: "" })),
    ).toBeUndefined();
  });
});
